 # Datentypen und Genauigkeit
#### Marcel Lüthi, Patrick Schnider<br/>Departement Mathematik und Informatik, Universität Basel

In diesem Notebook schauen wir uns vier grundlegende Datentypen von Java an. Sie lernen:
- den Wertebereich der Datentypen.
- wie Sie die Datentypen zwischeneinander konvertieren können.
- was Sie beachten müssen wenn es um die Genauigkeit geht.

### Datentypen für ganze Zahlen und Gleitkommazahlen

Sie haben den Unterschied zwischen ganzen Zahlen und Gleitkommazahlen in Java kennen gelernt. Dafür haben Sie bisher `int` und `double` verwendet. In Java gibt es aber noch den Datentpen `long` für ganze Zahlen, dessen Wertebereich grösser ist als der von `int`. Bei den Gleitkommazahlen gibt es noch den Datentypen `float`, welche die Zahlen weniger genau darstellt als `double`.

Entscheidungskriterien
- Ganze Zahlen `int`. Verwenden Sie `long` wenn der Wertebereich von `int` nicht gross genug ist.
- Gleitkommazahlen `double`. Verwenden Sie`float` wenn der Speicher zu begrenzt ist.
- (Nicht in der Vorlesung `byte` und `short`, welche ganze Zahlen nur in einem kleineren Bereich als `int` darstellen können.)

Um einen Zahlwert als `long` zu kennzeichnen, folgt nach der ganzen Zahl der Buchstabe `l` (kleines `L`).
```java
1l
9876543210l
```
Um einen Zahlenwert als `float` zu kennzeichnen, folgt nach dem Zahlwert der Buchstabe `f`.
```java
1f
-2.345f
3.1415e-3f
```

#### Miniübung
- Initialisieren Sie je eine Variable von Typ `long` und `float`.
- Weisen Sie die Variable vom Typ `float` einer Variable vom Typ `double` zu.
- Weisen Sie die Variable vom Typ `long` einer Variable vom Typ `int` zu.
- Was beobachten Sie?

In [ ]:
public class Values {
    public static void main(String[] args) {
        // Ihre Lösung kommt hier hin
        long l = 1l;
        float f = 1.0f;
        double d = f;
        // Nachfolgend reklamiert der Compiler das die verlangte Umwandlung
        // nicht einfach so erlaubt ist, da Genauigkeit verloren gehen könnte.
        int i = l;
    }
}

In [ ]:
Values.main(new String[0]);

#### Ordnung der Datentypen

Auf den Wertebereiche der Datentypen bezogen gilt:
$$ int \subset long \subset float \subset double$$

In Java ist die Zuweisung eines Wertes an eine Variable des gleichen oder *mächtigeren* Datentyps erlaubt.

#### Wertebereich

In Java gibt es definierte Konstanten welche den Wertebereich der jeweiligen Datentypen angeben.

Vorsicht: Die Konstanten mit dem selben Namen können für unterschiedliche Typen etwas anderes bedeuten. Dies könnte ein Hinweis auf nicht so guten Programmierstil sein.

Konstanten welche den Wertebereich angeben sind:

| Typ | Konstante | Bedeutung |
| --- | --- | --- |
| int | Integer.MAX_VALUE | grösster, darstellbarer Wert |
| int | Integer.MIN_VALUE | kleinster, darstellbarer Wert |
| long | Long.MAX_VALUE | grösster, darstellbarer Wert |
| long | Integer.MIN_VALUE | kleinster, darstellbarer Wert |
| float | Float.MAX_VALUE | grösster, darstellbarer Wert |
| float | Float.MIN_VALUE | kleinster, positiver, darstellbarer Wert |
| double | Double.MAX_VALUE | grösster, darstellbarer Wert |
| double | Double.MIN_VALUE | kleinster, positiver, darstellbarer Wert |

<div class="alert alert-block alert-info">
    Das Minimum der darstellbaren Gleitkommazahlen finden Sie als das negierte Maximum.
</div>

In [1]:
public class Range {
    public static void main(String[] args) {
        System.out.println("int stellt Zahlen von " +
            Integer.MIN_VALUE + " bis " +
            Integer.MAX_VALUE + " dar.");
        System.out.println("Dies ist in etwa +- 2 Milliarden.");
        System.out.println("long stellt Zahlen von " +
            Long.MIN_VALUE + " bis " +
            Long.MAX_VALUE + " dar.");
        System.out.println("Dies ist in etwa +- 9 Trillionen.");
        System.out.println("float stellt Zahlen von " +
            (-Float.MAX_VALUE) + " bis " +
            Float.MAX_VALUE + " dar.");
        System.out.println("double stellt Zahlen von " +
            (-Double.MAX_VALUE) + " bis " +
            Double.MAX_VALUE + " dar.");
        
        System.out.println(Float.MIN_VALUE + " ist nur etwas grösser als 0.0f.");
        System.out.println(Double.MIN_VALUE + " ist nur etwas grösser als 0.0.");
    }
}

In [2]:
Range.main(new String[0]);

int stellt Zahlen von -2147483648 bis 2147483647 dar.
Dies ist in etwa +- 2 Milliarden.
long stellt Zahlen von -9223372036854775808 bis 9223372036854775807 dar.
Dies ist in etwa +- 9 Trillionen.
float stellt Zahlen von -3.4028235E38 bis 3.4028235E38 dar.
double stellt Zahlen von -1.7976931348623157E308 bis 1.7976931348623157E308 dar.
1.4E-45 ist nur etwas grösser als 0.0f.
4.9E-324 ist nur etwas grösser als 0.0.


#### Welches ist der richtige Typ?

Welchen der vier Typen würden Sie bei den folgenden Aufgabenstellungen verwenden?
- "Speichern Sie den Geldbetrag eines Bankkunden."
- "Speichern Sie das Gewicht einer Person."
- "Zählen Sie die Häuser in der Stadt Basel."
- "Speichern Sie die Eulersche Zahl."
- "Zählen Sie alle Menschen auf dieser Erde."

Lösungsvorschlag:
- Geld -> am ehesten `long` für Rappenbeträge
- Gewicht -> `float` wohl genau genug
- Häuser -> `int` reicht im Moment noch gut
- Euler'sche Zahl -> möglichst genau, also `double`
- Menschen -> für `int` zu viele, deshalb `long`

Für diesen Kurs gilt in der Regel: Wenn nichts spezielles steht, nehmen Sie entweder `double` oder `int`. Dabei verwenden wir `double` für Messgrössen und `int` für Zählgrössen.

#### Speicherbedarf

Der Speicherbedarf wird gemessen, indem man die zum Speichern benötigten Bits, also die Anzahl Nullen und Einsen, angibt. Den Speicherbedarf können Sie für `int`, `long`, `float` und `double` mit Hilfe von in Java definierten Konstanten heraus finden.

Die Konstanten welche den Speicherbedarf angeben sind:

| Typ | Konstante |
| --- | --- |
| int | Integer.SIZE |
| long | Long.SIZE |
| float | Float.SIZE |
| double | Double.SIZE |


In [3]:
public class MemoryFootprint {
    public static void main(String[] args) {
        System.out.println("int belegt " + Integer.SIZE + " Bits im Speicher.");
        System.out.println("long belegt " + Long.SIZE + " Bits im Speicher.");
        System.out.println("float belegt " + Float.SIZE + " Bits im Speicher.");
        System.out.println("double belegt " + Double.SIZE + " Bits im Speicher.");
    }
}

In [4]:
MemoryFootprint.main(new String[0]);

int belegt 32 Bits im Speicher.
long belegt 64 Bits im Speicher.
float belegt 32 Bits im Speicher.
double belegt 64 Bits im Speicher.


### Konstanten

Variablen die ihren Wert nicht ändern sollen können mit dem Schlüsselwort `final` gekennzeichnet werden.

Beispiele:
- physikalische oder mathematische Konstanten
- unveränderbare Werte in einem Programm (z.B. Serveradresse, Programmname)

Namenskonvention: `ALL_CAPITAL_LETTERS`

In [8]:
public class ConstantValues {
    public static void main(String[] args) {
        final int MAX_LENGTH = 12;
        System.out.println(MAX_LENGTH + " und " + Math.PI + " sind Werte von Konstanten.");
        MAX_LENGTH = 20; // Kompiler reklamiert, da eine Wertzuweisung an Konstanten nicht erlaubt ist
    }
}

In [6]:
ConstantValues.main(new String[0]);

12 und 3.141592653589793 sind Werte von Konstanten.


##### Miniübung
- Versuchen Sie der Konstanten in der Zelle oben einen neuen Wert zuzuweisen.

### Rechenaufgabe

Berechnen Sie die Lösung für die folgende Aufgabe und geben Sie die Zahl aus.
- Sie wollen einen Gartenzaun der Länge 20m bauen.
- Ein Element hat die Länge 1.48m.
- Geben Sie aus wieviele Elemente Sie kaufen müssen.

In [18]:
public class GardenFence {
    public static void main(String[] args) {
        // Ihre Lösung kommt hier hin
        int length = 20;
        float element = 1.48f;
        double pieces = Math.ceil(length/element); // aufrunden, da nur ganze Elemente gekauft werden können
        System.out.println("Wir brauchen " + pieces + " Elemente."); // Ausgabe unbefriedigend, da 14.0 und nicht 14
    }
}

In [19]:
GardenFence.main(new String[0]);

Wir brauchen 14.0 Elemente.


### Type-Casting

Beim Type-Casting schreibt man den Typ der Variablen in runden Klammern vor den Ausdruck der umgewandelt werden soll.

Achtung: Das Umwandeln mit Type-Casting kommt meist auch mit einer gewissen Ungenauigkeit. Nachkommastellen werden einfach abgeschnitten.

Schema
```
(<TYPE>) <AUSDRUCK>
```

Beispiele
```java
(int) 2.0
(float) 3.1415
(long) (5.6 * 2.1)
```

#### Miniübung

- Übernehmen Sie die Berechnung aus der letzten Miniaufgabe.
- Erweitern Sie Ihren Code so, dass die Stückangabe als Variable vom Typ `int` berechnet und dann ausgegeben wird.

In [16]:
public class TypeCasts {
    public static void main(String[] args) {
        // Ihre Lösung kommt hier hin
        int length = 20;
        float element = 1.48f;
        int pieces = (int) Math.ceil(length/element);
        System.out.println("Wir brauchen " + pieces + " Elemente.");
    }
}

In [17]:
TypeCasts.main(new String[0]);

Wir brauchen 14 Elemente.


### Genauigkeit von Zahlen

Die Genauigkeit von Zahlen im Computer ist Beschränkt. Nicht nur beim Type-Casting können so Fehler entstehen. Wir schauen uns nun drei Rechnungen an wo nur die Addition verwendet wird. Werden wir die erwarteten Resultate erhalten.

#### Miniübung

Addieren Sie dreimal die Zahl 0.1 zueinander und geben das Resultat auf die Konsole aus.

In [20]:
public class Summation {
    public static void main(String[] args) {
        // Ihre Lösung kommt hier hin
        System.out.println(0.1 + 0.1 + 0.1);
    }
}

In [21]:
Summation.main(new String[0]);

0.30000000000000004


Mathematisch gibt die folgende Formel für positive, ganze Zahlen immer 1:
$$\sum\limits_{i=1}^{N}\frac{1}{N}$$

In [17]:
public class Summation2{
    public static void main(String[] args) {
        double result = 1.0 / 6.0 + 1.0 / 6.0 + 1.0 / 6.0 +
                        1.0 / 6.0 + 1.0 / 6.0 + 1.0 / 6.0;
        System.out.println("Erwartet 1.0, tatsächlich " + result);
    }
}

In [18]:
Summation2.main(new String[0]);

Erwartet 1.0, tatsächlich 0.9999999999999999


Dies zeigt, der Computer rechnet dies aus und löst die Aufgabe nicht irgendwie algebraisch.

Nachfolgend sehen Sie ein Beispiel einer Aufgabe in einem Lernsetting. Es geht darum drei Zahlen zu Summieren.

In [19]:
public class Summation3 {
    public static void main(String[] args) {
        double a = 13.0;
        double b = Math.PI;
        double c = 21.45;

        double solution_student = a + b + c;
        System.out.println("The solution of the student is: " + solution_student);

        double solution_teacher = a + c + b;
        System.out.println("The solution of the teacher is: " + solution_teacher);
    }
}

In [20]:
Summation3.main(new String[0]);

The solution of the student is: 37.59159265358979
The solution of the teacher is: 37.5915926535898


Diese Beispiele zeigen, dass man bei Gleitkommazahlen immer aufpassen muss mit der Genauigkeit.

Dies wird später im Kurs vorallem auch beim Testen von Programmen erneut wichtig.

#### Diskussion

- Zahlen vom Typ `float` werden mit 32 Bits dargestellt. Mit 32 Bits lassen sich $2^{32}$ unterschiedliche Werte darstellen.
- Der Wertebereich ist ungefähr $-3.4 \cdot 10^{38}$ bis $3.4 \cdot 10^{38}$, das ist ein Intervall der Grösse $2^{162}$.
- Intuition: In der wissenschaftlichen Schreibweise von Zahlen (z.B.: $3.14159\cdot10^{7}$), hat `float` eine Genauigkeit von 6 oder 7 signifikanten Dezimalstellen, für `double` sind es 14 bis 15 signifikante Dezimalstellen.
- Der absolute Abstand zweier benachbarter, darstellbaren Zahlen rund um 0.0 ist kleiner als für Zahlen nahe der Grenzen des darstellbaren Wertebereichs.

Interessierte finden eine ausführlichere und detailliertere Diskussion rund um die Darstellung und Genauigkeit von `float` [hier](https://introcs.cs.princeton.edu/java/91float/).

### Take Home

- Zwei Arten von Zahlen mit unterschiedlichen Zielen und Genauigkeit.
- Hierarchie von Zahlen: `int`, `long`, `float`, `double`
- Type Casting zum Umwandeln von Werten auf Kosten der Genauigkeit.
- Rechnungen mit den kennen gelernten Typen sind nicht immer exakt.